### PR TITLE
ENG-11129 shutdown save

### PR DESCRIPTION
**Pull request by vtkstef**
pro [pull request](https://github.com/VoltDB/pro/pull/1519)

### CODE REVIEW

File: src/frontend/org/voltdb/sysprocs/Pause.java [hunk 1]
```diff
@@ -54,14 +61,36 @@ public DependencyPair executePlanFragment(
...
+                        stat = zk.setData(VoltZK.operationMode,
+                                OperationMode.PAUSED.getBytes(), stat.getVersion());
+                    } catch (BadVersionException ex) {
+                        code = ex.code();
+                    }
+                } while (zkMode != OperationMode.PAUSED && code != Code.BADVERSION);
+
+                m_stat = stat;
+                voltdb.setMode(OperationMode.PAUSED);
+                voltdb.getHostMessenger().pause();
```
Comment (akhanzode) [inline src/frontend/org/voltdb/sysprocs/Pause.java]: Shouldnt hostmessenger pause before setMode?

File: src/frontend/org/voltdb/sysprocs/saverestore/NativeSnapshotWritePlan.java [hunk 2]
```diff
@@ -342,5 +342,9 @@ static void createFileBasedCompletionTasks(
         if (completionTask != null) {
             SnapshotSiteProcessor.m_tasksOnSnapshotCompletion.offer(completionTask);
         }
+        if (extraSnapshotData.getTerminus() != 0L) {
+            completionTask = SnapshotUtil.writeTerminusMarker(file_nonce, context.getPaths(), SNAP_LOG);
+            SnapshotSiteProcessor.m_tasksOnSnapshotCompletion.offer(completionTask);
```
Comment (akhanzode) [inline src/frontend/org/voltdb/sysprocs/saverestore/NativeSnapshotWritePlan.java]: Who gets the result of this task and handles failures?

File: tests/frontend/org/voltdb/regressionsuites/TestPrepareShutdown.java [hunk 3]
```diff
@@ -120,6 +124,107 @@ public void testPrepareShutdown() throws Exception {
...
+
+    static File getSnapshotPathForHost(LocalCluster cluster) {
+        if (cluster.isNewCli()) {
+            return new File(cluster.getServerSpecificRoot("0"), "snapshots");
+        } else {
+            List<File> subRoots = cluster.getSubRoots();
+            return new File (subRoots.get(0), "/tmp/" + System.getProperty("user.name") + "/snapshots");
+        }
+    }
+
```
Comment (akhanzode) [inline tests/frontend/org/voltdb/regressionsuites/TestPrepareShutdown.java]: TestPrepareShutdown should test multi node cluster. Its using hostcount as 1. This test will not be sufficient.

File: lib/python/voltcli/voltadmin.d/shutdown.py [hunk 4]
```diff
@@ -22,34 +22,40 @@
...
+            if resp.status() != 1:
+                runner.abort('The preparation for shutdown failed with status: %d' % resp.response.statusString)
+            snapsigil = resp.table(0).tuple(0).column_integer(0)
             runner.info('The cluster is paused prior to shutdown.')
             runner.info('Writing out all queued export data')
             status = runner.call_proc('@Quiesce', [], []).table(0).tuple(0).column_integer(0)
             if status <> 0:
-                runner.error('The cluster has failed to be quiesce with status: %d' % status)
-                return
+                runner.abort('The cluster has failed to be quiesce with status: %d' % status)
```
Comment (vtkstef) [inline lib/python/voltcli/voltadmin.d/shutdown.py]: elaborate please?

File: lib/python/voltcli/voltadmin.d/shutdown.py [hunk 5]
```diff
@@ -22,34 +22,40 @@
...
             runner.info('Completing outstanding export and DR transactions...')
             checkstats.check_export_dr(runner)
             runner.info('Completing outstanding client transactions.')
             checkstats.check_clients(runner)
             runner.info('Completing outstanding importer requests.')
             checkstats.check_importer(runner)
             runner.info('Cluster is ready for shutdown')
+            if runner.opts.save:
+                columns = [VOLT.FastSerializer.VOLTTYPE_BIGINT]
+                shutdown_params =  [snapsigil]
```
Comment (vtkstef) [inline lib/python/voltcli/voltadmin.d/shutdown.py]: it is a sigil. maybe you want to call it zk_pause_txnid?

File: src/frontend/org/voltdb/ExtensibleSnapshotDigestData.java [hunk 6]
```diff
@@ -52,13 +53,21 @@
      */
     private final Map<Integer, JSONObject> m_drMixedClusterSizeConsumerState;
 
+    /**
+     * Value that denotes that this snapshot is one created with shutdown --save. 0
+     * being no, and other values yes
+     */
+    private long m_terminus;
```
Comment (vtkstef) [inline src/frontend/org/voltdb/ExtensibleSnapshotDigestData.java]: nope they get merged when multiple write attempts are merged in. They attempt a write, if they get get a bad version, they read it from zk and merge the nodes values into it, and attempt a new write

File: src/frontend/org/voltdb/InvocationDispatcher.java [hunk 7]
```diff
@@ -408,18 +420,26 @@ else if ("@SnapshotRestore".equals(procName)) {
...
                     return useSnapshotCatalogToRestoreSnapshotSchema(task, handler, ccxn, user);
                 }
+            } else if ("@Shutdown".equals(procName)) {
+                if (task.getParams().size() == 1) {
+                    return takeShutdownSaveSnapshot(task, handler, ccxn, user);
+                }
+            }
+            // Verify that admin mode sysprocs are called from a client on the
+            // admin port, otherwise return a failure
+            if ((   "@Pause".equals(procName)
```
Comment (vtkstef) [inline src/frontend/org/voltdb/InvocationDispatcher.java]: elaborate please?

File: src/frontend/org/voltdb/InvocationDispatcher.java [hunk 8]
```diff
@@ -922,6 +943,187 @@ private final static void transmitResponseMessage(ClientResponse r, Connection c
...
+    private final ClientResponseImpl takeShutdownSaveSnapshot(
+            final StoredProcedureInvocation task,
+            final InvocationClientHandler handler, final Connection ccxn,
+            final AuthUser user
+            )
+    {
+        // shutdown save snapshot is available for Pro edition only
+        if (!MiscUtils.isPro()) {
+            task.setParams();
+            dispatch(task, handler, ccxn, user);
```
Comment (vtkstef) [inline src/frontend/org/voltdb/InvocationDispatcher.java]: good catch

File: src/frontend/org/voltdb/InvocationDispatcher.java [hunk 9]
```diff
@@ -922,6 +943,187 @@ private final static void transmitResponseMessage(ClientResponse r, Connection c
...
+                    .key(SnapshotUtil.JSON_URIPATH).value("file://" + paths.resolve(paths.getSnapshoth()).getPath())
+                    .key(SnapshotUtil.JSON_NONCE).value(SnapshotUtil.getShutdownSaveNonce(zkTxnId))
+                    .key(SnapshotUtil.JSON_TERMINUS).value(zkTxnId)
+                    .key(SnapshotUtil.JSON_BLOCK).value(true)
+                    .key(SnapshotUtil.JSON_PATH_TYPE).value(SnapshotPathType.SNAP_AUTO.toString())
+                    .key(SnapshotUtil.JSON_FORMAT).value(SnapshotFormat.NATIVE.toString())
+                    .endObject();
+            snapshotJson = jss.toString();
+        } catch (JSONException e) {
+            VoltDB.crashLocalVoltDB("Failed to create shutdown snapshot save command", true, e);
```
Comment (vtkstef) [inline src/frontend/org/voltdb/InvocationDispatcher.java]: we usually crash if we cannot serialize into json

File: src/frontend/org/voltdb/sysprocs/saverestore/NativeSnapshotWritePlan.java (hunk 2 above)
Comment (vtkstef) [inline src/frontend/org/voltdb/sysprocs/saverestore/NativeSnapshotWritePlan.java]: elaborate please. these tasks are offered, their status are reported on when they are executed at snapshot completion

File: lib/python/voltcli/voltadmin.d/shutdown.py (hunk 5 above)
Comment (akhanzode) [inline lib/python/voltcli/voltadmin.d/shutdown.py]: yes zk_pause_txnid please.

File: src/frontend/org/voltdb/ExtensibleSnapshotDigestData.java [hunk 10]
```diff
@@ -144,6 +153,12 @@ private void mergeExportSequenceNumbersToZK(JSONObject jsonObj, VoltLogger log)
         }
     }
 
+    private void mergeTerminusToZK(JSONObject jsonObj) throws JSONException {
+        long jsTerminus = jsonObj.optLong(SnapshotUtil.JSON_TERMINUS, 0L);
+        m_terminus = Math.max(jsTerminus, m_terminus);
+        jsonObj.put(SnapshotUtil.JSON_TERMINUS, m_terminus);
+    }
+
     private void writeDRTupleStreamInfoToSnapshot(JSONStringer stringer) throws IOException {
```
Comment (akhanzode) [inline src/frontend/org/voltdb/ExtensibleSnapshotDigestData.java]: This method does not look like DR specific.

File: src/frontend/org/voltdb/sysprocs/saverestore/NativeSnapshotWritePlan.java (hunk 2 above)
Comment (akhanzode) [inline src/frontend/org/voltdb/sysprocs/saverestore/NativeSnapshotWritePlan.java]: What I am saying is if we failed to write magic marker what happens is not clear.

**Pull request merged**